### DISCUSSION TITLE
Indexeddb support in Quasar, specifically for use outside .vue files

### DISCUSSION THREAD

**Discussion by FKPSC** [quasarframework/quasar#15327] [Ideas / Proposals] [answered]
There was a discussion here: https://github.com/quasarframework/quasar/issues/580

It seems like the idea had enough approval from everyone but it just wasn't the right time.

It is also the final missing piece of offline storage after localStorage and sessionStorage.

I don't know the inner workings of such functionality, but I am thinking it could be added as one of both of the following:

- Bring the functionality of IndexedDB to same level as localStorage and sessionStorage so it can be used in and oustside of .vue files. If compatible, it would allow using any existing IndexedDB library (localForage, Dexie.js, JsStore, MiniMongo, PouchDB, idb, idb-keyval, $mol_db, RxDB,... ) .
- Quasar could have its own wrapper, ie its own library, like the ones mentioned above to provide basic db functionality and even extended functionality as it has been done with the type preservation for localStorage and sessionStorage.

If this is a lost harder than localStorage and sessionStorage, it would be nice to know if it is possible or have a long-term goal for it (maybe Quasar v3 :smile:  )

**Comment by TobyMosque** [comment, marked as answer]
The Storage wrappers provided by the quasar are minimal on design, if u need complex needs, there is recommended to look somewhere else.

u can use dexie, lokijs, SQLite (capacitor/Electron) or any other database library in your project, all what u need is initialize that in a boot and inject in the pages/components. u can expose that globally, but I advocate u would use DI and composables everywhere.

Regarding the storage itself, maybe u wanna to look into the storages composables, so u'll get a reactive storage.

https://vueuse.org/core/usestorage/

And if u like pinia, u can leverage the storage by using the pinia persisted state plugin:

https://prazdevs.github.io/pinia-plugin-persistedstate/

**Comment by FKPSC** [comment]
I actually found out that using Dexie works even in the Pinia. The only thing I have to keep in mind is that, since Dexie is async, I can only use in Pinia actions and not getters. That fits my usecase though.

I have not found a great tutorial/article with Quasar or Pinia being used with Dexie. That might be a good opportunity for someone with the know-how and a blog/YouTube channel :smile:
